File: HmsWalletServerDemo/src/main/java/com/huawei/wallet/hms/hmssdk/WalletBuildService.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.huawei.wallet.hms.hmssdk;

import com.huawei.wallet.hms.hmssdk.dto.HwWalletObject;

import java.util.List;

/**
 * Interface of invoking wallet api gateway service.
 *
 * @since 2019-11-05
 */
public interface WalletBuildService {
    /**
     * Create a HwWalletObject.
     *
     * @param requestParam JSON-type HwWalletObject.
     * @param instanceId instance ID.
     * @param modelId model ID.
     * @return the created HwWalletObject.
     */
    HwWalletObject createHwWalletObject(String requestParam, String instanceId, String modelId);

    /**
     * Post a HwWalletObject to wallet server.
     *
     * @param url request URL parameter.
     * @param hwWalletObject the HwWalletObject.
     * @return the posted HwWalletObject.
     */
    HwWalletObject postHwWalletObjectToWalletServer(String url, HwWalletObject hwWalletObject);

    /**
     * Return the wallet model for a given model ID.
     *
     * @param url request URL parameter.
     * @param modelId model ID.
     * @return the wallet model.
     */
    HwWalletObject getWalletModelByModelId(String url, String modelId);

    /**
     * Return a list of some wallet model created by a specific issuer.
     *
     * @param url request URL parameter.
     * @param pageSize maximum number of wallet model in the returned list. All wallet model created by the issuer
     *        will be returned if pageSize is null.
     * @return the wallet model list.
     */
    List<HwWalletObject> listWalletModel(String url, Integer pageSize);

    /**
     * Return the wallet instance for a given instance ID.
     *
     * @param url request URL parameter.
     * @param instanceId instance ID.
     * @return the wallet instance.
     */
    HwWalletObject getWalletInstanceByInstanceId(String url, String instanceId);

    /**
     * Return a list of some wallet instance belong to a specific wallet model.
     *
     * @param url request URL parameter.
     * @param modelId model ID.
     * @param pageSize max number of wallet instance in the returned list. All wallet instance belongs to the wallet
     *        model will be returned if pageSize is null.
     * @return the wallet instance list.
     */
    List<HwWalletObject> listWalletInstance(String url, String modelId, Integer pageSize);

    /**
     * Overwrite a wallet model.
     *
     * @param url request URL parameter.
     * @param modelId model ID.
     * @param body JSON-type HwWalletObject.
     * @return the updated wallet model.
     */
    HwWalletObject fullUpdateWalletModel(String url, String modelId, String body);

    /**
     * Update a wallet model.
     *
     * @param url request URL parameter.
     * @param modelId model ID.
     * @param body JSON-type HwWalletObject.
     * @return the updated wallet model.
     */
    HwWalletObject partialUpdateWalletModel(String url, String modelId, String body);

    /**
     * Overwrite a wallet instance.
     *
     * @param url request URL parameter.
     * @param instanceId instance ID.
     * @param body JSON-type wallet instance.
     * @return the updated wallet instance.
     */
    HwWalletObject fullUpdateWalletInstance(String url, String instanceId, HwWalletObject body);

    /**
     * Update a wallet instance.
     *
     * @param url request URL parameter.
     * @param instanceId instance ID.
     * @param body JSON-type wallet instance.
     * @return the updated wallet instance.
     */
    HwWalletObject partialUpdateWalletInstance(String url, String instanceId, HwWalletObject body);

    /**
     * Add messages to a HwWalletObject.
     *
     * @param url request URL parameter.
     * @param resourceId model ID or instance ID.
     * @param body JSON-type messageList object.
     * @return the updated HwWalletObject.
     */
    HwWalletObject addMessageToHwWalletObject(String url, String resourceId, String body);

    /**
     * Add/remove linked offer IDs to/from a loyalty instance.
     *
     * @param url request URL parameter.
     * @param instanceId instance ID.
     * @param body JSON-type LinkedOfferInstanceIds object.
     * @return the updated loyalty instance.
     */
    HwWalletObject updateLinkedOffersToLoyaltyInstance(String url, String instanceId, String body);
}

File: wallet/hms/wallet_build_service.py
# Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# -*- coding: utf-8 -*-
# @Date     : 2020/6/4
# @version  : Python 3.8.3
# @File     : wallet_build_service
import json
import requests
from wallet.hms.dto.batch_get_hw_wallet_result import BatchGetHwWalletResult
from wallet.hms.dto.hw_wallet_object import HwWalletObject
from wallet.hms.dto.page_info import PageInfo
from wallet.util.common_util import dict_to_obj, isNoneOrEmptyString, isNoneObject, str_to_dict
from wallet.util.config_helper import ConfigHelper
from wallet.util.open_gateway_client import get_token

CONFIG_HELPER = ConfigHelper()


def postHwWalletObjectToWalletServer(url, hwWalletObject):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    wallet_server_url = baseUrl + url
    # Send the http request and get response.
    rsp = requests.post(wallet_server_url,
                        data=json.dumps(hwWalletObject, default=lambda obj: obj.__dict__, sort_keys=True, indent=4),
                        headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def getHwWalletObjectById(url, resourceId):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    walletServerUrl = baseUrl + url + resourceId
    # Send the http request and get response.
    rsp = requests.get(walletServerUrl, headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    # Return the HwWalletObject with the corresponding resourceId.
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def getModels(url, pageSize):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    walletServerUrl = baseUrl + url
    if pageSize is None:
        response = requests.get(walletServerUrl, headers=headers)
        batchGetHwWalletResult: BatchGetHwWalletResult = dict_to_obj(response.text, BatchGetHwWalletResult())
        if isNoneObject(batchGetHwWalletResult):
            print("listWalletModel failed. batchGetHwWalletResult is null.")
            return None
        else:
            return batchGetHwWalletResult.data
    else:
        walletModelList = []
        tempUrl = walletServerUrl + "?pageSize=" + str(pageSize)
        session = ""
        while True:
            finalUrl = tempUrl
            if not isNoneOrEmptyString(session):
                finalUrl = finalUrl + "&session=" + session
            response = requests.get(finalUrl, headers=headers)
            batchGetHwWalletResult: BatchGetHwWalletResult = dict_to_obj(str_to_dict(response.text),
                                                                         BatchGetHwWalletResult())
            if isNoneObject(batchGetHwWalletResult):
                print("listWalletModel failed. batchGetHwWalletResult is null.")
                return None
            data = batchGetHwWalletResult.get_data()
            if data is None:
                return walletModelList
            walletModelList.extend(data)
            pageInfo: PageInfo = dict_to_obj(batchGetHwWalletResult.get_pageInfo(), PageInfo())
            if isNoneObject(pageInfo):
                print("listWalletModel failed. pageInfo is null.")
                return None
            next_session = pageInfo.get_nextSession()
            if isNoneOrEmptyString(next_session):
                return walletModelList
            session = next_session


def getInstances(url, modelId, pageSize):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl') + url
    if pageSize is None:
        response = requests.get(url + "?modelId=" + modelId, headers=headers)
        batchGetHwWalletResult: BatchGetHwWalletResult = dict_to_obj(str_to_dict(response.text),
                                                                     BatchGetHwWalletResult())
        if isNoneObject(batchGetHwWalletResult):
            print("listWalletInstance failed. batchGetHwWalletResult is null.")
            return None
        else:
            return batchGetHwWalletResult.data
    else:
        walletInstanceList = []
        tempUrl = baseUrl + "?modelId=" + modelId + "&pageSize=" + str(pageSize)
        session = ""
        while True:
            finalUrl = tempUrl
            if not isNoneOrEmptyString(session):
                finalUrl = finalUrl + "&session=" + session
            response = requests.get(finalUrl, headers=headers)
            batchGetHwWalletResult: BatchGetHwWalletResult = dict_to_obj(str_to_dict(response.text),
                                                                         BatchGetHwWalletResult())
            if isNoneObject(batchGetHwWalletResult):
                print("listWalletInstance failed. batchGetHwWalletResult is null.")
                return None
            data = batchGetHwWalletResult.get_data()
            if data is None:
                return walletInstanceList
            walletInstanceList.extend(data)
            pageInfo: PageInfo = dict_to_obj(batchGetHwWalletResult.get_pageInfo(), PageInfo())
            if isNoneObject(pageInfo):
                print("listWalletInstance failed. pageInfo is null.")
                return None
            next_session = pageInfo.get_nextSession()
            if isNoneOrEmptyString(next_session):
                return walletInstanceList
            session = next_session


def fullUpdateHwWalletObject(url, resourceId, hwWalletObject):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    walletServerUrl = baseUrl + url + resourceId
    # Send the http request and get response.
    rsp = requests.put(walletServerUrl,
                       data=json.dumps(hwWalletObject, default=lambda obj: obj.__dict__, sort_keys=True, indent=4),
                       headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    # Return the updated wallet object
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def partialUpdateHwWalletObject(url, resourceId, hwWalletObject):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    walletServerUrl = baseUrl + url + resourceId
    # Send the http request and get response.
    rsp = requests.patch(walletServerUrl,
                         data=json.dumps(hwWalletObject, default=lambda obj: obj.__dict__, sort_keys=True, indent=4),
                         headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    # Return the updated wallet model.
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def addMessageToHwWalletObject(url, resourceId, body):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    wallet_server_url = baseUrl + url.replace("addMessage", "") + resourceId + "/addMessage"
    rsp = requests.post(wallet_server_url,
                        data=json.dumps(body, default=lambda obj: obj.__dict__, sort_keys=True, indent=4),
                        headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def updateLinkedOffersToLoyaltyInstance(url, resourceId, body):
    # Construct the http header.
    headers = construct_http_headers()
    # Construct the http URL.
    baseUrl = CONFIG_HELPER.get_value('walletServerBaseUrl')
    wallet_server_url = baseUrl + url.replace("linkedoffers", resourceId) + "/linkedoffers"
    rsp = requests.patch(wallet_server_url,
                         data=json.dumps(body, default=lambda obj: obj.__dict__, sort_keys=True, indent=4),
                         headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    return dict_to_obj(str_to_dict(rsp.text), HwWalletObject())


def postToWalletServer(url_segment, body):
    headers = construct_http_headers()
    base_url = CONFIG_HELPER.get_value('walletServerBaseUrl')
    wallet_server_url = base_url + url_segment
    rsp = requests.post(wallet_server_url, data=body, headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    return rsp.text


def delete_from_wallet_server(url_segment):
    headers = construct_http_headers()
    base_url = CONFIG_HELPER.get_value('walletServerBaseUrl')
    wallet_server_url = base_url + url_segment
    rsp = requests.delete(wallet_server_url, headers=headers)
    if rsp.status_code != requests.codes.ok:
        rsp.raise_for_status()
    return rsp.text

def construct_http_headers():
    client_id = CONFIG_HELPER.get_value('gw.appid')
    client_secret = CONFIG_HELPER.get_value('gw.appid.secret')
    authorization = get_token(client_id, client_secret)
    return {'Content-Type': 'application/json;charset=utf-8', 'Authorization': 'Bearer ' + authorization,
            'Accept': 'application/json;charset=utf-8'}
